extensions/amp-live-list/0.1/poller.js

```diff
@@ -0,0 +1,186 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+import {exponentialBackoff} from '../../../src/exponential-backoff';
+
+
+/**
+ * Polling state of a Poller instance.
+ * @enum {string}
+ */
+export const PollerState = {
+  /**
+   * The poller is running at a the set interval.
+   */
+  RUNNING: 'running',
+
+  /**
+   * The poller is running but is doing exponential backoff (using
+   * the original interval value as base) every tick.
+   */
+  BACKEDOFF: 'backedoff',
+
+  /**
+   * The poller is not running and not executing any work.
+   */
+  STOPPED: 'stopped',
+};
+
+
+/**
+ * Poller with backoff functionality.
+ */
+export class Poller {
+
+  constructor(win, wait, work) {
+    /** @private @const {!Window} */
+    this.win = win;
+
+    /** @private {number} */
+    this.wait_ = wait;
+
+    /** @private {function(): !Promise} */
+    this.work_ = work;
+
+    /** @private {?number} */
+    this.lastTimeoutId_ = null;
+
+    /** @private {?number} */
+    this.lastBackoffTimeoutId_ = null;
+
+    /** @private {function(): number} */
+    this.backoffPoller_ = this.getBackoffPoller_();
+
+    /** @private {!PollerState} */
+    this.state_ = PollerState.STOPPED;
+
+    /**
+     * Mostly for testing purposes.
+     * @private {!Promise}
+     */
+    this.lastWorkPromise_ = null;
+  }
+
+  /**
+   * @return {function(): number}
+   * @private
+   */
+  getBackoffPoller_() {
+    return exponentialBackoff();
+  }
+
+  /**
+   * @return {!PollerState}
+   */
+  getState() {
+    return this.state_;
+  }
+
+  /**
+   * Initalize any work needed to start polling.
+   */
+  start() {
+    if (this.lastTimeoutId_) {
+      return;
+    }
+    if (this.lastBackoffTimeoutId_) {
+      this.win.clearTimeout(this.lastBackoffTimeoutId_);
+    }
+
+    // we reset the backoff poller
+    this.backoffPoller_ = this.getBackoffPoller_();
+    this.state_ = PollerState.RUNNING;
+    this.poll_();
+  }
+
+  /**
+   * Initialize backoff function and set correct state.
+   */
+  backoff() {
```

Comment:
so i've removed all the `catch` code blocks in the poller and still have `if (err.retriable)` code in the live-list-manager, if you feel strongly that the logic should be here in the poller let me know. i wanted to keep the poller as stupid as possible